### ISSUE TITLE
Memcache for the Homepage

### ISSUE THREAD

**Issue by teolemon** [openfoodfacts/openfoodfacts-server#8792]
Loading the webpage takes two seconds. That means, but we probably don't use memcache to serve it. Amirite ?
@stephanegigandet @raphael0202

**Comment by raphael0202** [comment]
caching in memcache is enabled, but it looks like there is an error when saving MongoDB results in memcached:

```
[267909] /opt/product-opener/lib/ProductOpener/Display.pm 1375 mongodb {} get_cache_results - miss - key: openfoodfacts.net:search_products/32d75fc9c3e8b63dd2f640a563e199ee
[268135] /opt/product-opener/lib/ProductOpener/Display.pm 1390 mongodb {} set_cache_results - setting value - key: openfoodfacts.net:search_products/32d75fc9c3e8b63dd2f640a563e199ee - total_size: 26584379
[269054] /opt/product-opener/lib/ProductOpener/Display.pm 1398 mongodb {} set_cache_results - error - key: openfoodfacts.net:search_products/32d75fc9c3e8b63dd2f640a563e199ee
```

the test was performed on staging. On prod when I filtered `mongodb_log4perl.log` by 'error', errors were logged everytime I hit the home page (non logged-in user), so I suspect the issue can be reproduced in staging.  

it works correctly locally.

**Comment by teolemon** [comment]
Nice catch @raphael0202 👍

**Comment by raphael0202** [comment]
After further investigation, here is what I found.
Due to personalized search, we always request all fields to MongoDB when searching products on OFF: https://github.com/openfoodfacts/openfoodfacts-server/blob/main/lib/ProductOpener/Display.pm#L4978
Most popular products (as displayed on the main page of OFF) are the most complete ones (and the ones with the most data), hence the giant 30MB response from MongoDB.
We should definitely restrict the fields to only the required ones. We can expect from this:
- faster queries to MongoDB
- use of memcached (faster loading time)

**Comment by stephanegigandet** [comment]
> We should definitely restrict the fields to only the required ones.

Definitely. We'll have to be very careful to list all the fields we need for the attributes.

Also there are some fields we may need to organize differently (e.g. images, so that we don't pull the list of the 542324 images of nutella just to know which one is the front image...)

**Comment by raphael0202** [comment]
> Definitely. We'll have to be very careful to list all the fields we need for the attributes.

Is it listed somewhere? I had trouble understanding where this is computed in Product Opener code.

I have done a quick analysis on a popular product (Nutella, ) of the largest subfields (as measured by serializing the subfield content in JSON):

```
# Full obj: 114 KB

# Subfields of size >= 1KB

# popularity_tags: 43 KB (1600 string items)
# images: 23 KB
# ecoscore_data: 6.5 KB
# source: 3.5 KB
# editor_tags: 2.8 KB
# nutriments: 1.8KB
# correctors_tags: 1.7 KB
# ingredients: 1.5 KB
# misc_tags: 1.32 KB
# ecoscore_extended_data: 1.23 KB

# remaining fields: 34 KB
```

`popularity_tags` is an easy target, but we need each object to be <= 10KB if we want to respect the 1MB limit (as we store 100 results), so it's obviously not enough.
Memcached max object size can be increased, but it's not adviced by Memcached documentation.

**Comment by teolemon** [comment]
editor_tags, corrector_tags, misc_tags, ecoscore_extended_data, source (and popularity_tags) are not currently used for attributes I believe.
Sidenote: And the more time passes, the more I believe the Search project might bring truly personal search at scale

**Comment by stephanegigandet** [comment]
@raphael0202 the code for computing attributes is in lib/ProductOpener/Attributes.pm

To get some of the fields needed:

~/openfoodfacts-server$ grep 'product_ref->' lib/ProductOpener/Attributes.pm 
~/openfoodfacts-server$ grep 'has_tag' lib/ProductOpener/Attributes.pm 

Something stupid I did 12 years ago: "images" contains both all the uploaded images + the selected images. We only care about selected images. At some point we should separate them. It's an issue for the popular products that tends to have tons of images.

A workaround could be to do it at runtime for caching: we still get all images from mongo, but we remove the ones we don't need before caching. We probably can do the same for some other fields.
